File: solutions/java/phone-number/1/src/main/java/PhoneNumber.java
// For my Shree DR.MDD

class PhoneNumber {
    private final String sanitizedNumber;

    PhoneNumber(String inputNumber) {
        if (inputNumber.matches(".*[a-zA-Z].*")) throw new IllegalArgumentException("letters not permitted");
        if (inputNumber.matches(".*[,!?;:\"'\\[\\]{}<>].*")) throw new IllegalArgumentException("punctuations not permitted");

        inputNumber = inputNumber.replaceAll("[^0-9]", "");
        final int numLength = inputNumber.length();

        if (numLength < 10) throw new IllegalArgumentException("must not be fewer than 10 digits");
        if (numLength > 11) throw new IllegalArgumentException("must not be greater than 11 digits");

        if (numLength == 11) {
            if (inputNumber.charAt(0) == '1') inputNumber = inputNumber.substring(1);
            else throw new IllegalArgumentException("11 digits must start with 1");
        }

        String invalidCodeType = null;
        String invalidCodeDigit = null;
        int codeValue = Character.getNumericValue(inputNumber.charAt(0));
        if (codeValue < 2) invalidCodeType = "area";
        else {
            codeValue = Character.getNumericValue(inputNumber.charAt(3));
            if (codeValue < 2) invalidCodeType = "exchange";
        }
        invalidCodeDigit = codeValue == 0 ? "zero" : "one";
        if (invalidCodeType != null) throw new IllegalArgumentException(invalidCodeType + " code cannot start with " + invalidCodeDigit);

        sanitizedNumber = inputNumber;
    }

    String getNumber() {
        return sanitizedNumber;
    }
}

File: solutions/ocaml/phone-number/1/phone_number.ml
(* Dedicated to Junko F. Didi and Shree DR.MDD *)

let filter s =
  s
  |> String.to_seq
  |> List.of_seq
  |> List.fold_left
       (fun acc ch ->
         match ch with
         | '0' .. '9' -> (int_of_char ch - 48) :: acc
         | _ -> acc)
       []
  |> List.rev

let is_valid digits =
  if List.hd digits = 0 then
    Error "area code cannot start with zero"
  else if List.hd digits = 1 then
    Error "area code cannot start with one"
  else if List.nth digits 3 = 0 then
    Error "exchange code cannot start with zero"
  else if List.nth digits 3 = 1 then
    Error "exchange code cannot start with one"
  else
    Ok
      (digits
       |> List.map (( + ) 48)
       |> List.map char_of_int
       |> List.to_seq
       |> String.of_seq)

let number s =
  let rec scan chars =
    match chars with
    | [] ->
        let seq = filter s in
        if List.length seq > 11 then
          Error "must not be greater than 11 digits"
        else if List.length seq = 11 && List.hd seq <> 1 then
          Error "11 digits must start with 1"
        else if List.length seq = 11 && List.hd seq = 1 then
          is_valid (List.tl seq)
        else if List.length seq = 10 then
          is_valid seq
        else
          Error "must not be fewer than 10 digits"
    | hd :: tl -> (
        match hd with
        | 'A' .. 'Z' | 'a' .. 'z' -> Error "letters not permitted"
        | '@' | ':' | '!' -> Error "punctuations not permitted"
        | _ -> scan tl)
  in
  scan (s |> String.to_seq |> List.of_seq)
